### PR TITLE
Armada Gamemode Update

### PR DESCRIPTION
**Pull request by Karmic-Skink**
## About The Pull Request
This PR contains an update to the currently offline Armada gamemode.
It contains:
- 4 new objectives
- New fleets
- New ships
- Backend Jank

Note: Some aspects remain placeholders for testing purposes
Also intentionally disabled from pool without admin intervention

## Why It's Good For The Game
Fleshing out additional gamemodes potentially adds more depth to the standard loop by giving crews specific goals to achieve.

## Changelog
:cl:
tweak: Updated the Armada gamemode
add: Added four new objectives
add: Added new fleets
add: Added new elite ships
code: A number of backend changes to signals and other small things
/:cl:

### CODE REVIEW

File: nsv13/code/game/gamemodes/overmap/objectives/rescue_fleet.dm [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+	status = 2
+
+/datum/overmap_objective/rescue_fleet/proc/encounter_success()
+	if(status != 0)
+		return
+
+	priority_announce("These resources should prove invaluable in our conflict with the Syndicate, good work.")
+	status = 1
+	var/datum/star_system/S = SSstar_system.system_by_id("Staging")
+	var/datum/fleet/N = locate(/datum/fleet/nanotrasen/convoy) in selected_system
```
Comment (TheFakeElon) [inline nsv13/code/game/gamemodes/overmap/objectives/rescue_fleet.dm:58]:
```suggestion
	var/datum/fleet/nanotrasen/convoy/N = locate() in selected_system
```

File: nsv13/code/game/gamemodes/overmap/objectives/system_defence_armada.dm [hunk 2]
```diff
@@ -20,6 +20,12 @@
...
 
 	var/datum/star_system/target = SSstar_system.system_by_id(selected_system)
-	if(target.alignment != "nanotrasen")
+	var/obj/structure/overmap/OM = SSstar_system.find_main_overmap()
+	if(target.alignment != "nanotrasen" && OM.current_system != target)
 		priority_announce("Attention [station_name()]. Our presence in this sector has been severely diminished due to your incompetence. Return to base immediately for disciplinary action.", "Naval Command")
+		status = 2
 		SSovermap_mode.mode.defeat()
+
+	if(target.enemies_in_system.len == 0)
```
Comment (TheFakeElon) [inline nsv13/code/game/gamemodes/overmap/objectives/system_defence_armada.dm:29]: Generally speaking, you should use `length()` when reading length and `.len` when modifying it. length() is faster
```suggestion
	if(!length(target.enemies_in_system))
```

File: nsv13/code/modules/overmap/traders.dm [hunk 3]
```diff
@@ -223,6 +203,45 @@
...
+////OBJECTIVE SPECIFIC
+/datum/trader/damaged_cruiser
+	name = "Damaged Caracal Class Cruiser" //TEMP
+	desc = "Nanotrasen Caracal Class Cruiser" //TEMP
+	shortname = "NTC"
+	faction_type = FACTION_ID_NT
+	sold_items = null //Nothing for sale
+	greetings = list(
+		"We've got our hands fully currently, what did you need?",\
+		"I don't mean to be terse, but we are busy over here.",\
```
Comment (TheFakeElon) [inline nsv13/code/modules/overmap/traders.dm:215]: You don't need to escape newlines in a list
```suggestion
		"We've got our hands fully currently, what did you need?",
		"I don't mean to be terse, but we are busy over here.",
```

File: nsv13/code/modules/overmap/types/syndicate.dm [hunk 4]
```diff
@@ -351,6 +351,16 @@
 	weapon_types[FIRE_MODE_GAUSS] = new /datum/ship_weapon/gauss(src) //AI ships want to be able to use gauss too. I say let them...
 	weapon_types[FIRE_MODE_MISSILE] = new/datum/ship_weapon/missile_launcher(src)
 
+/obj/structure/overmap/syndicate/ai/assault_cruiser/boarding_frigate/elite
+	name = "Special Ops marine frigate"
```
Comment (TheFakeElon) [inline nsv13/code/modules/overmap/types/syndicate.dm:355]:
```suggestion
	name = "special ops marine frigate"
```

Comment (TheFakeElon) [reply nsv13/code/modules/overmap/types/syndicate.dm:355]: Unless "special ops" is the name of a specific group, in which case just use /improper

File: nsv13/code/modules/overmap/types/syndicate.dm [hunk 5]
```diff
@@ -381,6 +391,18 @@
 	weapon_types[FIRE_MODE_GAUSS] = new /datum/ship_weapon/gauss(src) //AI ships want to be able to use gauss too. I say let them...
 	weapon_types[FIRE_MODE_MISSILE] = new/datum/ship_weapon/missile_launcher(src)
 
+/obj/structure/overmap/syndicate/ai/gunboat/elite
+	name = "Special Ops anti-air frigate"
```
Comment (TheFakeElon) [inline nsv13/code/modules/overmap/types/syndicate.dm:395]: ditto

File: nsv13/code/modules/overmap/types/syndicate.dm [hunk 6]
```diff
@@ -415,6 +437,19 @@
 	weapon_types[FIRE_MODE_GAUSS] = new /datum/ship_weapon/gauss(src) //AI ships want to be able to use gauss too. I say let them...
 	weapon_types[FIRE_MODE_MISSILE] = new/datum/ship_weapon/missile_launcher(src)
 
+/obj/structure/overmap/syndicate/ai/submarine/elite
+	name = "Special Ops Sub-spacemarine"
```
Comment (TheFakeElon) [inline nsv13/code/modules/overmap/types/syndicate.dm:441]: ditto

File: nsv13/code/modules/overmap/traders_items.dm [hunk 7]
```diff
@@ -39,13 +39,41 @@
 	faction = "syndicate"
 	supply_pod_type = /obj/structure/closet/supplypod/syndicate_odst
 
+/obj/structure/overmap/trader/damaged_cruiser
+	name = "Damaged Cruiser"
```
Comment (TheFakeElon) [inline nsv13/code/modules/overmap/traders_items.dm:43]:
```suggestion
	name = "damaged cruiser"
```

File: nsv13/code/modules/overmap/traders_items.dm [hunk 8]
```diff
@@ -39,13 +39,41 @@
...
+	name = "NT Staging Post" //TEMP
+	obj_integrity = 7500 //Really need this to last for a while
+	max_integrity = 7500
+	ai_controlled = TRUE
+
+/obj/structure/overmap/trader/nt_staging_post/apply_weapons()
+	weapon_types[FIRE_MODE_PDC] = new /datum/ship_weapon/pdc_mount(src)
+
+/obj/structure/overmap/trader/syndi_listening_post
+	name = "Syndicate Listening Post"
```
Comment (TheFakeElon) [inline nsv13/code/modules/overmap/traders_items.dm:57]:
```suggestion
	name = "\improper Syndicate listening post"
```

File: nsv13/code/modules/overmap/ai-skynet.dm [hunk 9]
```diff
@@ -327,81 +328,81 @@ Adding tasks is easy! Just define a datum for it.
...
-			// Able to check off multiple objectives through the loop if crew are piling everything into one torpedo 
-			receipt.completed_objectives += objective 
-			expecting_cargo -= request 
-			
-			// Break from the loop if there are multiple cargo missions requesting the same type of item. No double dipping! 
-			break 
+		return FALSE
+
+	for ( var/datum/overmap_objective/cargo/request in expecting_cargo ) // Only validate this station's cargo related objectives
+		if ( request.status != 0 )
```
Comment (TheFakeElon) [inline nsv13/code/modules/overmap/ai-skynet.dm:382]:
```suggestion
		if (request.status)
```

File: nsv13/code/modules/overmap/ai-skynet.dm [hunk 10]
```diff
@@ -327,81 +328,81 @@ Adding tasks is easy! Just define a datum for it.
...
+			continue
+
+		var/datum/overmap_objective/cargo/objective = request
```
Comment (TheFakeElon) [inline nsv13/code/modules/overmap/ai-skynet.dm:385]: This doesn't do anything, it repaths `request` to the path is already has

File: nsv13/code/modules/overmap/ai-skynet.dm [hunk 11]
```diff
@@ -707,15 +717,26 @@ Adding tasks is easy! Just define a datum for it.
...
 	name = "\proper Syndicate Armada" //Fleet spawned if the players are too inactive. Set course...FOR EARTH.
 	destroyer_types = list(/obj/structure/overmap/syndicate/ai, /obj/structure/overmap/syndicate/ai/nuclear, /obj/structure/overmap/syndicate/ai/assault_cruiser, /obj/structure/overmap/syndicate/ai/gunboat, /obj/structure/overmap/syndicate/ai/submarine, /obj/structure/overmap/syndicate/ai/assault_cruiser/boarding_frigate)
 	size = FLEET_DIFFICULTY_VERY_HARD
 	allow_difficulty_scaling = FALSE
 	taunts = list("We're coming for Sol, and you can't stop us. All batteries fire at will.", "Lay down your arms now, you're outnumbered.", "All hands, assume assault formation. Begin bombardment.")
 	audio_cues = list()
 
-/datum/fleet/interdiction	//Pretty strong fleet with unerring hunting senses, Adminspawn for now.
+/datum/fleet/armada
+	name = "\proper Syndicate Armada" //Fleet used in the armada gamemode for the final confrontation
```
Comment (TheFakeElon) [inline nsv13/code/modules/overmap/ai-skynet.dm:729]: Names starting with a capital letter a proper by default

Comment (TheFakeElon) [reply nsv13/code/modules/overmap/ai-skynet.dm:729]: Also are you sure you want it to be proper? 
proper name reference = "It's Syndicate Armada"
improper name reference = "It's the Syndicate Armada"

Generally you only want it for pure names
```suggestion
	name = "\improper Syndicate Armada" //Fleet used in the armada gamemode for the final confrontation
```

File: nsv13/code/game/gamemodes/overmap/objectives/defend_station.dm [hunk 12]
```diff
@@ -0,0 +1,90 @@
...
+	F.current_system = selected_system
+	F.assemble(selected_system)
+
+	if(waves >= 1)
+		addtimer(CALLBACK(src, .proc/send_fleet), 3 MINUTES) //Small interval between waves
+
+	if(waves < 1)
+		last_wave = TRUE //Unlock possible victory
+
+/datum/overmap_objective/defend_station/proc/station_destroyed()
```
Comment (TheFakeElon) [inline nsv13/code/game/gamemodes/overmap/objectives/defend_station.dm:80]:
```suggestion
/datum/overmap_objective/defend_station/proc/station_destroyed()
	SIGNAL_HANDLER
```

File: nsv13/code/game/gamemodes/overmap/objectives/defend_station.dm [hunk 13]
```diff
@@ -0,0 +1,90 @@
...
+	status = 2
+	priority_announce("Reports indicate that staging post's reactor went critical, search for survivors if possible and continue your primary objective.")
+
+/datum/overmap_objective/defend_station/check_completion()
+	if(!last_wave)
+		return
+
+	if(selected_system.enemies_in_system.len == 0)
```
Comment (TheFakeElon) [inline nsv13/code/game/gamemodes/overmap/objectives/defend_station.dm:88]:
```suggestion
	if(!length(selected_system.enemies_in_system))
```

File: nsv13/code/game/gamemodes/overmap/objectives/rescue_fleet.dm [hunk 14]
```diff
@@ -0,0 +1,59 @@
...
+
+/datum/overmap_objective/rescue_fleet/proc/encounter_failure()
+	if(status != 0)
+		return
+
+	priority_announce("We've lost contact with the transport convoy, search for survivors if possible and continue your primary objective.")
+	status = 2
+
+/datum/overmap_objective/rescue_fleet/proc/encounter_success()
+	if(status != 0)
```
Comment (TheFakeElon) [inline nsv13/code/game/gamemodes/overmap/objectives/rescue_fleet.dm:52]:
```suggestion
	if(status)
```

**Review by TheFakeElon — COMMENTED** [review]
obj_integrity defaults to max_integrity, so you don't need to define each one individually.

**Comment by Karmic-Skink** [comment]
I'll address code cleanup once this has been proof of concept tested.

File: nsv13/code/game/gamemodes/overmap/objectives/defend_station.dm (hunk 12 above)
Comment (Karmic-Skink) [reply nsv13/code/game/gamemodes/overmap/objectives/defend_station.dm:80]: ?

**Comment by Karmic-Skink** [comment]
No apparent interest in this code.

**Pull request closed without merging**